File: src/core/account/login-cookies.ts
/**
 * 从 WebView 的 cookie 串里认出登录凭证（API 文档 §0.2）。
 *
 * 登录成功的标志是同时出现 `ngaPassportUid` 与 `ngaPassportCid` 两个 Cookie；
 * 登录**前**页面就会挂着占位值（uid 可能是 `guest`、cid 可能是短垃圾串），
 * 所以光看键存在不够，还得按取值形状过滤，否则轮询会在登录前误报。
 */

/** 解析 `k=v; k2=v2` 形式的 cookie 串。值里允许再出现 `=`（base64 之类）。 */
export function parseCookieString(cookie: string): ReadonlyMap<string, string> {
  const jar = new Map<string, string>()
  for (const part of cookie.split(';')) {
    const eq = part.indexOf('=')
    if (eq < 0) continue
    const key = part.slice(0, eq).trim()
    const value = part.slice(eq + 1).trim()
    if (key !== '') jar.set(key, value)
  }
  return jar
}

export interface LoginCookies {
  readonly uid: string
  /** 会话凭证 ngaPassportCid */
  readonly cid: string
  /** 原始 ngaPassportUrlencodedUname（GBK 双重 URLEncode），可能缺失 */
  readonly urlencodedUname: string | null
}

/** uid 必须是纯数字——排掉 `guest` 这类未登录占位。 */
const UID_PATTERN = /^\d+$/
/** cid 是长随机串（实测 40 位字母数字）；短值/空值视为还没登录完。 */
const CID_PATTERN = /^[0-9A-Za-z_-]{16,}$/

/** cookie 串里有合法凭证就取出来，否则 null（表示登录还没完成）。 */
export function extractLoginCookies(cookie: string): LoginCookies | null {
  const jar = parseCookieString(cookie)
  const uid = jar.get('ngaPassportUid') ?? ''
  const cid = jar.get('ngaPassportCid') ?? ''
  if (!UID_PATTERN.test(uid) || !CID_PATTERN.test(cid)) return null
  return { uid, cid, urlencodedUname: jar.get('ngaPassportUrlencodedUname') ?? null }
}

File: app/src/main/kotlin/com/chasel/ng2n/data/account/LoginCookies.kt
package com.chasel.ng2n.data.account

import com.chasel.ng2n.core.net.encoding.REPLACEMENT_CHAR
import com.chasel.ng2n.core.net.encoding.decodeGb18030

fun parseCookieString(cookie: String): Map<String, String> {
  val jar = LinkedHashMap<String, String>()
  for (part in cookie.split(';')) {
    val eq = part.indexOf('=')
    if (eq < 0) continue
    val key = part.substring(0, eq).trim()
    val value = part.substring(eq + 1).trim()
    if (key != "") jar[key] = value
  }
  return jar
}

data class LoginCookies(
  val uid: String,
  val cid: String,
  val urlencodedUname: String?,
)

private val UID_PATTERN = Regex("^\\d+$")

private val CID_PATTERN = Regex("^[0-9A-Za-z_-]{16,}$")

fun extractLoginCookies(cookie: String): LoginCookies? {
  val jar = parseCookieString(cookie)
  val uid = jar[COOKIE_PASSPORT_UID_NAME] ?: ""
  val cid = jar[COOKIE_PASSPORT_CID_NAME] ?: ""
  if (!UID_PATTERN.matches(uid) || !CID_PATTERN.matches(cid)) return null
  return LoginCookies(uid = uid, cid = cid, urlencodedUname = jar[COOKIE_UNAME_NAME])
}

private const val COOKIE_PASSPORT_UID_NAME = "ngaPassportUid"
private const val COOKIE_PASSPORT_CID_NAME = "ngaPassportCid"
private const val COOKIE_UNAME_NAME = "ngaPassportUrlencodedUname"

fun decodeLoginUsername(raw: String): String? {
  val once = percentDecodeToBytes(raw) ?: return null
  val asText = buildString(once.size) { for (byte in once) append((byte.toInt() and 0xff).toChar()) }
  val twice = percentDecodeToBytes(asText) ?: return null
  val name = decodeGb18030(twice).trim()
  if (name == "" || name.contains(REPLACEMENT_CHAR)) return null
  return name
}

private fun percentDecodeToBytes(input: String): ByteArray? {
  val bytes = ArrayList<Byte>(input.length)
  var index = 0
  while (index < input.length) {
    val char = input[index]
    if (char == '%') {
      if (index + 3 > input.length) return null
      val high = hexDigit(input[index + 1]) ?: return null
      val low = hexDigit(input[index + 2]) ?: return null
      bytes += ((high shl 4) or low).toByte()
      index += 3
      continue
    }
    if (char == '+') {
      bytes += 0x20.toByte()
      index += 1
      continue
    }
    if (char.code > 0xff) return null
    bytes += char.code.toByte()
    index += 1
  }
  return bytes.toByteArray()
}

private fun hexDigit(char: Char): Int? = when (char) {
  in '0'..'9' -> char - '0'
  in 'a'..'f' -> char - 'a' + 10
  in 'A'..'F' -> char - 'A' + 10
  else -> null
}
